include/common/aglResCommon.h

```diff
@@ -0,0 +1,198 @@
+#pragma once
+
+#include <basis/seadRawPrint.h>
+#include <prim/seadBitUtil.h>
+
+namespace agl {
+
+// maybe first parameter is is_le, maybe it is big_endian - different between decomps
+void ModifyEndianU32(bool is_le, void* p_data, size_t size);
+
+template <typename _DataType>
+class ResCommon {
+public:
+    using DataType = _DataType;
+
+    ResCommon() : mpData(nullptr) {}
+
+    ResCommon(const void* data) : mpData(static_cast<const DataType*>(data)) {}
+
+    bool isValid() const { return mpData != nullptr; }
+
+    void verify() const {
+        if (isValidMagic()) {
+            const char* b = reinterpret_cast<const char*>(mpData);
+            SEAD_ASSERT_MSG(false, "Wrong binary. [%c%c%c%c].", b[0], b[1], b[2], b[3]);
+        }
+
+        if (isValidVersion()) {
+            SEAD_ASSERT_MSG(false, "Version error.current:%d binary:%d",
+                        DataType::getVersion(), sead::BitUtil::bitCastPtr<u32>(ptr(), 4));
+        }
+    }
+
+    DataType* ptr() {
+        assertValid();
+        return const_cast<DataType*>(mpData);
+    }
+
+    const DataType* ptr() const {
+        assertValid();
+        return mpData;
+    }
+
+    u8* ptrBytes() const { return const_cast<u8*>(reinterpret_cast<const u8*>(mpData)); }
+
+    DataType& ref() {
+        SEAD_ASSERT(isValid());
+        return *ptr();
+    }
+
+    const DataType& ref() const {
+        SEAD_ASSERT(isValid());
+        return *ptr();
+    }
+
+    bool isValidMagic() const {
+        return sead::BitUtil::bitCastPtr<u32>(ptr(), 0) == DataType::getSignature();
+    }
+
+    bool isValidVersion() const {
+        return sead::BitUtil::bitCastPtr<u32>(ptr(), 4) == DataType::getVersion();
+    }
+
+    void assertValid() const { SEAD_ASSERT(isValid()); }
+
+protected:
+    const DataType* mpData;
+};
+
+#define AGL_RES_COMMON(class_name)                                                                 \
```

Comment:
This seems unnecessary. You can just do

```
using ResCommon::ResCommon;
```

in the derived classes to inherit from the base ResCommon constructor